crates/ruff_linter/src/rules/ruff/rules/post_init_default.rs

```diff
@@ -0,0 +1,187 @@
+use anyhow::Context;
+
+use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_semantic::{Scope, ScopeKind};
+use ruff_python_trivia::{indentation_at_offset, textwrap};
+use ruff_text_size::Ranged;
+
+use crate::{checkers::ast::Checker, importer::ImportRequest};
+
+use super::helpers::is_dataclass;
+
+/// ## What it does
+/// Checks for `__post_init__` dataclass methods with parameter defaults.
+///
+/// ## Why is this bad?
+/// Adding a default value to a parameter in a `__post_init__` method has no
+/// impact on whether the parameter will have a default value in the dataclass's
+/// generated `__init__` method. To create an init-only dataclass parameter with
+/// a default value, you should use an `InitVar` field in the dataclass's class
+/// body and give that `InitVar` field a default value.
+///
+/// As the [documentation] states:
+///
+/// > Init-only fields are added as parameters to the generated `__init__()`
+/// > method, and are passed to the optional `__post_init__()` method. They are
+/// > not otherwise used by dataclasses.
+///
+/// ## Example
+/// ```python
+/// from dataclasses import InitVar, dataclass
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: InitVar[int] = 0
+///
+///     def __post_init__(self, bar: int = 1, baz: int = 2) -> None:
+///         print(bar, baz)
+///
+///
+/// foo = Foo()  # Prints '0 2'.
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from dataclasses import InitVar, dataclass
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: InitVar[int] = 1
+///     baz: InitVar[int] = 2
+///
+///     def __post_init__(self, bar: int, baz: int) -> None:
+///         print(bar, baz)
+///
+///
+/// foo = Foo()  # Prints '1 2'.
+/// ```
+///
+/// ## References
+/// - [Python documentation: Post-init processing](https://docs.python.org/3/library/dataclasses.html#post-init-processing)
+/// - [Python documentation: Init-only variables](https://docs.python.org/3/library/dataclasses.html#init-only-variables)
+///
+/// [documentation]: https://docs.python.org/3/library/dataclasses.html#init-only-variables
+#[violation]
+pub struct PostInitDefault;
+
+impl Violation for PostInitDefault {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__post_init__` method with argument defaults")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(format!("Use `dataclasses.InitVar` instead"))
+    }
+}
+
+/// RUF033
+pub(crate) fn post_init_default(checker: &mut Checker, function_def: &ast::StmtFunctionDef) {
```

Comment:
I made this run on `StmtFunctionDef` nodes rather than `StmtClassDef` nodes, as this allows us to access `checker.semantic().current_scope()` to see what other symbols have been bound in the class. This enables us to see much more simply and more accurately whether it's safe to add a new `InitVar` field to the class body in the fix